File: web-common/src/java/org/jepria/web/ssr/table/Table.java
package org.jepria.web.ssr.table;


import java.util.List;
import java.util.Objects;
import java.util.Set;

import org.jepria.web.ssr.El;
import org.jepria.web.ssr.Text;

public abstract class Table<T extends ItemData> extends El {
  
  private int tabIndexValue;
  
  public static interface TabIndex {
    /**
     * Assigns the next {@code tabindex} to the element
     * @param el
     */
    void setNext(El el);
  }
  
  protected final Text text;
  
  public Table(Text text) {
    super("div");
    this.text = text;
    
    // TODO remove id, use class and css
    setAttribute("id", "table");
    setAttribute("style", "width: 100%;");
    
    
    addScript("js/table.js");
    addStyle("css/table.css");
    addStyle("css/common.css"); // for .field-text
  }
  
  protected boolean isEditable() {
    return true;
  }
  
  /**
   * Clear the table and load new items 
   * 
   * @param items basic table items, possibly modified
   * @param itemsCreated optional items to render {@code .created} table rows, may be null
   * @param itemsDeleted optional item (set of {@link Dto#id}) to render {@code .deleted} table rows, may be null
   */
  public void load(List<T> items, List<T> itemsCreated, Set<String> itemsDeleted) {
    
    // reset
    childs.clear();
    tabIndexValue = 1;
    
    
    final TabIndex tabIndex = new TabIndex() {
      @Override
      public void setNext(El el) {
        el.setAttribute("tabindex", tabIndexValue++);
      }
    };
    
    
    appendChild(createHeader());
    
    if (items != null) {
      
      boolean evenOddGray = true; // for unmodifiable table
      
      for (T item: items) {
        final String itemId = item.getId();
        
        final El row = createRow(item, tabIndex);
        
        if (!isEditable()) {
          if (evenOddGray) {
            row.addClass("even-odd-gray");
          }
          evenOddGray = !evenOddGray;
        }
        
        if (itemsDeleted != null && itemsDeleted.contains(itemId)) {
          row.classList.add("deleted");
        } else {
          // check any field modified
          if (item.values().stream().anyMatch(
              field -> !Objects.equals(field.value, field.valueOriginal))) {
            row.classList.add("modified");
          }
        }
        
        row.setAttribute("item-id", item.getId());
        
        appendChild(row);
      }
    }
    if (itemsCreated != null) {
      for (T item: itemsCreated) {
        final El row = createRowCreated(item, tabIndex);
        row.setAttribute("item-id", item.getId());
        appendChild(row);
      }
    }
    
    setAttribute("tabindex-next", tabIndexValue);
  }
  
  /**
   * Create a basic table row (possibly modified) representing a single item
   * @param item data from the server, non-null
   * @param tabIndex table-wide counter for assigning {@code tabindex} attributes to {@code input} elements
   * @return
   */
  public abstract El createRow(T item, TabIndex tabIndex);
  
  /**
   * Creates a table row (in created state) representing a single item, that has not been saved to the server yet,
   * but having a UI table row created for saving
   * @param item optional data from the UI to fill the created table row with, may be null
   * @param tabIndex table-wide counter for assigning {@code tabindex} attributes to {@code input} elements
   * @return
   */
  public abstract El createRowCreated(T item, TabIndex tabIndex);
  
  protected El addField(El cell, Field field, String placeholder) {
    return addField(cell, field, placeholder, isEditable());
  }
  
  protected El addField(El cell, Field field, String placeholder, boolean fieldEditable) {
    final El fieldEl;
    if (fieldEditable) {
      fieldEl = createFieldInput(field.name, field.value, placeholder);
    } else {
      fieldEl = createFieldLabel(field.value);
    }
    
    return addField(cell, field, fieldEl, placeholder);
  }
  
  protected El addField(El cell, Field field, El fieldEl, String placeholder) {
    
    fieldEl.classList.add("field-text");
    fieldEl.classList.add("field-text_inactivatible");
    fieldEl.classList.add("disableable");
    
    if (field.readonly) {
      fieldEl.setReadonly(true);
    } else {
      fieldEl.setAttribute("value-original", field.valueOriginal);
    }
    
    if (field.invalid) {
      fieldEl.classList.add("invalid");
      if (field.invalidMessage != null) {
        fieldEl.setAttribute("title", field.invalidMessage);
      }
    }
    
    El wrapper = wrapCellPad(fieldEl);
    cell.appendChild(wrapper);

    if (isEditable()) {
      addStrike(cell);
    }
      
    return fieldEl;
  }
  
  protected El createFieldInput(String name, String value, String placeholder) {
    El field = new El("input");
    field.setAttribute("type", "text");
    field.setAttribute("name", name);
    field.setAttribute("value", value);
    field.setAttribute("placeholder", placeholder);
    return field;
  }

  protected void addStrike(El cell) {
    El strike = new El("div");
    strike.classList.add("strike");
    cell.appendChild(strike);
  }

  protected El createFieldLabel(String value) {
    El field = new El("label");
    field.setInnerHTML(value, true);
    return field;
  }

  protected El wrapCellPad(El element) {
    El wrapper = new El("div");
    
    El leftDiv = new El("div");
    leftDiv.classList.add("cell-pad-left");
    wrapper.appendChild(leftDiv);
    
    El rightDiv = new El("div");
    rightDiv.classList.add("cell-pad-right");
    wrapper.appendChild(rightDiv);
    
    El midDiv = new El("div");
    midDiv.classList.add("cell-pad-mid");
    wrapper.appendChild(midDiv);
    
    midDiv.appendChild(element);
    
    return wrapper;
  }
  
  protected void addFieldDelete(El cell, TabIndex tabIndex) {

    El field = new El("div");
    
    El buttonDelete = new El("input");
    buttonDelete.classList.add("button-delete");
    buttonDelete.classList.add("button-delete_delete");
    buttonDelete.setAttribute("type", "image");
    buttonDelete.setAttribute("src", "img/delete.png");
    buttonDelete.setAttribute("title", text.getString("org.jepria.web.ssr.table.buttonDelete.title.delete"));
    tabIndex.setNext(buttonDelete);
    
    El buttonUndelete = new El("input");
    buttonUndelete.classList.add("button-delete");
    buttonUndelete.classList.add("button-delete_undelete");
    buttonUndelete.setAttribute("type", "image");
    buttonUndelete.setAttribute("src", "img/undelete.png");
    buttonUndelete.setAttribute("title", text.getString("org.jepria.web.ssr.table.buttonDelete.title.undelete"));
    tabIndex.setNext(buttonUndelete);
    
    field.appendChild(buttonDelete);
    field.appendChild(buttonUndelete);
    
    El wrapper = wrapCellPad(field);  
    cell.appendChild(wrapper);
    
    //return field;
  }
  
  protected El createCell(El row, String columnClass) {
    El cell = new El("div");
    cell.classList.add("cell");
    cell.classList.add(columnClass);
    row.appendChild(cell);
    return cell;
  }
  
  protected CheckBox addCheckbox(El cell, Field field) {
    
    boolean active = !"false".equals(field.value);
    CheckBox checkbox = new CheckBox(active);
    
    checkbox.setEnabled(!field.readonly);

    if (field.readonly) {
      checkbox.setReadonly(true);
    } else {
      checkbox.setAttribute("value-original", !"false".equals(field.valueOriginal));
    }
    
    checkbox.classList.add("table__checkbox");
    
    checkbox.classList.add("disableable");

    if (field.invalid) {
      checkbox.classList.add("invalid");
      if (field.invalidMessage != null) {
        checkbox.setAttribute("title", field.invalidMessage);
      }
    }
    
    El wrapper = wrapCellPad(checkbox);  
    cell.appendChild(wrapper);
    
    if (isEditable()) {
      addStrike(cell);
    }
    
    // add text attributes
    checkbox.setAttribute("org.jepria.web.ssr.Table.checkbox_active.title.inactive", text.getString("org.jepria.web.ssr.Table.checkbox_active.title.inactive"));
    checkbox.setAttribute("org.jepria.web.ssr.Table.checkbox_active.title.active", text.getString("org.jepria.web.ssr.Table.checkbox_active.title.active"));
    
    
    return checkbox;
  }
  
  protected abstract El createHeader();
}

File: web-common/src/java/org/jepria/web/ssr/table/table.js
// table.js:
function table_onload() {
  var table = document.getElementById("table-container").firstElementChild;
  
  var fieldsText = table.querySelectorAll("input.field-text");
  for (var i = 0; i < fieldsText.length; i++) {
    var fieldText = fieldsText[i];
     
    fieldText.oninput = function(field) {// javascript doesn't use block scope for variables
      return function() {
        onFieldInput(field);
      
        // clear field 'invalid' state on manual value change only
        field.classList.remove("invalid");
        field.removeAttribute("title"); // TODO what if a field's title is not related to the invalid state?
      }
    }(fieldText);
  }
  
  var checkboxes = document.getElementsByClassName("table__checkbox");
  for (var i = 0; i < checkboxes.length; i++) {
    var checkbox = checkboxes[i];
    
    checkbox.onclick = function(checkbox) {// javascript doesn't use block scope for variables
      return function(event) {
        // JS note: the handler is being invoked twice per single click 
        // (first for onclick:event.target=SPAN, second for onclick:event.target=INPUT)
        if (event.target.tagName.toLowerCase() == "input") {
          onCheckboxInput(checkbox);
          
          // checkbox cannot have invalid value: either it is not modifiable at all, or the value is valid
        }
      }
    }(checkbox);
  }
  
  addFieldsDeleteScript(table);
  
  // disable deleted rows
  var deletedRows = table.querySelectorAll(".row.deleted");
  for (var i = 0; i < deletedRows.length; i++) {
    setDisabled(deletedRows[i], true);
  }
  
  triggerFieldsInput(table);
}

function onFieldInput(field) {
  onFieldValueChanged(field, field.value);
}

function onFieldValueChanged(field, newValue) {
  valueOriginal = field.getAttribute("value-original");
  if (typeof valueOriginal === 'undefined') {
    // treat as modified
    field.classList.add("modified");
  } else {
    if (valueOriginal !== newValue) {
      field.classList.add("modified");
    } else {
      field.classList.remove("modified");
    }
  }
  
  checkModifications();
}


/**
 * Checks for any user modifications throughout the table
 */
function checkModifications() {
  totalModifications = 
      document.getElementsByClassName("modified").length
      - document.querySelectorAll(".row.created.deleted .modified").length
      + document.getElementsByClassName("row deleted").length;
  
  var buttonEnabled = totalModifications > 0;
  setButtonSaveEnabled(buttonEnabled);
  setButtonResetEnabled(buttonEnabled);
}

/**
 * Disable everything '.deletable' in a composite
 * @param disabled
 * @returns
 */
function setDisabled(composite, disabled) {
  var disableableElements = composite.querySelectorAll("input.deletable, button.deletable");

  for (var i = 0; i < disableableElements.length; i++) {
    var disableableElement = disableableElements[i]; 
    disableableElement.disabled = disabled;
    // disableableElement.setAttribute("readonly", true); // alternative
  }
  
  var checkboxes = composite.querySelectorAll(".checkbox.deletable");
  for (var i = 0; i < checkboxes.length; i++) {
    var checkbox = checkboxes[i];
    if (!checkbox.classList.contains("readonly")) {
      setCheckboxEnabled(checkbox, !disabled);
    }
  }
}

function addFieldsDeleteScript(composite) {
  var fieldsDelete = composite.querySelectorAll("input.button-delete");
  for (var i = 0; i < fieldsDelete.length; i++) {
    fieldsDelete[i].onclick = function(event){onDeleteButtonClick(event.target)};
  }
}

function onDeleteButtonClick(button) {
  //TODO resolve the relative path:
  var row = button.parentElement.parentElement.parentElement.parentElement.parentElement;
  
  if (row.classList.contains("created")) {
    // for newly created rows just remove them from table
    row.parentNode.removeChild(row);
    
  } else if (!row.classList.contains("deleted")) {
    row.classList.add("deleted");

    setDisabled(row, true);
    
  } else {
    row.classList.remove("deleted");
    
    setDisabled(row, false);
  }
  
  checkModifications();
}

function onButtonCreateClick() {
  
  var table = document.getElementById("table");

  var newRowTemplate = document.getElementById("table-new-row-template-container").firstElementChild;
  var newRow = newRowTemplate.cloneNode(true);
  
  table.appendChild(newRow);
  
  addNewRowScript(newRow);
  
  
  //set 'tabindex' attributes by 'tabindex-rel' attribute
  var tabIndexAnchor = +table.getAttribute("tabindex-next");
  var maxTabIndexRel = 0;
  var hasTabIndexRels = Array.from(newRow.getElementsByClassName("has-tabindex-rel"));// element.getElementsByClassName FETCHES the elements 
  for (var i = 0; i < hasTabIndexRels.length; i++) {
    var hasTabIndexRel = hasTabIndexRels[i];
    var tabIndexRel = +hasTabIndexRel.getAttribute("tabindex-rel");
    maxTabIndexRel = Math.max(maxTabIndexRel, tabIndexRel);
    hasTabIndexRel.setAttribute("tabindex", tabIndexAnchor + tabIndexRel);
    
    // cleanup
    hasTabIndexRel.removeAttribute("tabindex-rel");
    hasTabIndexRel.classList.remove("has-tabindex-rel");
  }
  table.setAttribute("tabindex-next", tabIndexAnchor + maxTabIndexRel + 1);
  
  // TODO set proper tabindexes for control buttons
  
  
  // focus on the first text input field
  newRow.querySelectorAll(".cell input[type='text']")[0].focus();
  
  // scroll to the created row (bottom)
  window.scrollTo(0, document.body.scrollHeight);
  
  
  checkModifications();
}

function onButtonSaveClick() {
  
  var rowsModified = getRowsModified();
  var rowsDeleted = getRowsDeleted();
  var rowsCreated = getRowsCreated();
  
  modRequestList = [];
  
  if (rowsModified.length > 0) {
    for (var i = 0; i < rowsModified.length; i++) {
      modRequestList.push(
          {
            id: rowsModified[i].itemId, // same as id
            action: "update", 
            data: rowsModified[i].itemData
          }
      );
    }
  }
  
  if (rowsDeleted.length > 0) {
    for (var i = 0; i < rowsDeleted.length; i++) {
      modRequestList.push(
          {
            id: rowsDeleted[i], // same as id
            action: "delete"
          }
      );
    }
  }
  
  if (rowsCreated.length > 0) {
    for (var i = 0; i < rowsCreated.length; i++) {
      modRequestList.push(
          {
            id: "row-create-" + (i + 1),
            action: "create",
            data: rowsCreated[i].itemData
          }
      );
    }
  }
  
  if (modRequestList.length > 0) {
    xhttp = new XMLHttpRequest();
    xhttp.open("POST", getSsrUrlMod(), true); // TODO reference to the applicational function
    xhttp.onreadystatechange = function() {
      if (xhttp.readyState != 4) return;
      window.location.href = getSsrUrlBase(); // TODO reference to the applicational function
    }
    xhttp.send(JSON.stringify(modRequestList));
  } else {
    // TODO report nothing to save
  }
}

function getRowsModified() {
  var rows = document.querySelectorAll("#table div.row");
  var data = [];
  for (var i = 0; i < rows.length; i++) {
    row = rows[i];
    if (!row.classList.contains("deleted") && !row.classList.contains("created") && row.getElementsByClassName("modified").length > 0) {
      var rowData = collectRowData(row);
      data.push({itemId: row.getAttribute("item-id"), itemData: rowData});
    }
  }
  return data;
}

function getRowsDeleted() {
  var rows = document.querySelectorAll("#table div.row");
  rowsDeletedIds = [];
  for (var i = 0; i < rows.length; i++) {
    row = rows[i];
    if (row.classList.contains("deleted") && !row.classList.contains("created")) {
      rowsDeletedIds.push(row.getAttribute("item-id"));
    }
  }
  return rowsDeletedIds;
}

function getRowsCreated() {
  var rows = document.querySelectorAll("#table div.row");
  var data = [];
  for (var i = 0; i < rows.length; i++) {
    var row = rows[i];
    if (row.classList.contains("created") && !row.classList.contains("deleted")) {
      var rowData = collectRowData(row);
      data.push({itemId: row.getAttribute("item-id"), itemData: rowData});
    }
  }
  return data;
}

/**
 * Collect modified row fields
 * @param row to collect data from
 */
function collectRowData(row) {
  var data = {};
  
  var fields = row.querySelectorAll(".cell-field input.modified");
  
  for (var j = 0; j < fields.length; j++) {
    var field = fields[j];
    if (field.name === "active") {
      // TODO workaround. If checkbox becomes a class, make its own property 'value'
      data[field.name] = field.checked;
    } else {
      data[field.name] = field.value;
    }
  }
  
  return data;
}

function addNewRowScript(rowCreate) {
  addFieldsDeleteScript(rowCreate);
  triggerFieldsInput(rowCreate);
}

function triggerFieldsInput(composite) {
  var fields = composite.querySelectorAll("input.field-text");
  for (var i = 0; i < fields.length; i++) {
    onFieldInput(fields[i]);
  }
  
  var checkboxes = composite.querySelectorAll(".checkbox");
  for (var i = 0; i < checkboxes.length; i++) {
    onCheckboxInput(checkboxes[i]);
  }
}

function onButtonResetClick() {
  xhttp = new XMLHttpRequest();
  xhttp.open("POST", getSsrUrlReset(), true); // TODO reference to the applicational function
  xhttp.onreadystatechange = function() {
    if (xhttp.readyState != 4) return;
    window.location.href = getSsrUrlBase(); // TODO reference to the applicational function
  } 
  xhttp.send();
}
// :table.js